feat: improve auto-detection check for safetensors models

jaideepr97 opened the pull request:
<!-- Thank you for contributing to InstructLab! -->

<!-- STEPS TO FOLLOW:
  1. Add a description of the changes (frequently the same as the commit description)
  2. Enter the issue number next to "Resolves #" below (if there is no tracking issue resolved, **remove that section**)
  3. Follow the steps in the checklist below, starting with the **Commit Message Formatting**.
-->

Adds new `is_model_safetensors` function that checks if supplied model path is a directory, and if so, checks if it qualifies as a safetensors model. Since a valid safetensors model should have the safetensors itself, along with required files for the tokenizer, we check to ensure all these files are present in the directory before concluding that the supplied modle path is a valid safetensors model 

**Issue resolved by this Pull Request:**
Resolves #1457 

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

danmcp commented:
Good candidate for a unit test at some point.

leseb commented on src/instructlab/model/backends/backends.py:
```diff
@@ -121,12 +154,18 @@ def determine_backend(model_path: pathlib.Path) -> str:
     """
-    # Check if the model is a GGUF file
+    # If the model is a safetensors model and OS is linux, it's a vLLM model
+    try:
```
Is it correct to assume that does code should only run on Linux? If so, how about we wrap it around an `if sys.platform == "linux"` to avoid executing it on other platforms?

jaideepr97 replied:
pushed an update
IMO I think its better to check that the model is a valid safetensors file first, and then check for the OS so that we can catch the case where someone is trying to use a valid safetensor model on a different OS and put out a helpful message, as opposed to not running the check at all which might be confusing for the user

leseb commented on src/instructlab/model/backends/backends.py:
```diff
@@ -86,6 +88,37 @@ def shutdown(self):
+    """
+
+    if not model_path.is_dir():
+        return False
```
At this point we can't tell whether there is an issue with the content of the directory or if model_path was not a directory at all.

jaideepr97 replied:
removed, thanks!

leseb commented on src/instructlab/model/backends/backends.py:
```diff
@@ -121,12 +154,18 @@ def determine_backend(model_path: pathlib.Path) -> str:
     """
-    # Check if the model is a GGUF file
+    # If the model is a safetensors model and OS is linux, it's a vLLM model
+    try:
+        is_safetensors = is_model_safetensors(model_path)
+    except Exception as e:
+        raise ValueError(
+            f"Failed to determine whether the model is a safetensors model: {e}"
+        ) from e
```
```suggestion
if model_path.is_dir() and sys.platform == "linux":
    try:
        is_safetensors = is_model_safetensors(model_path)
    except Exception as e:
        raise ValueError(
            f"Failed to determine whether the model is a safetensors model: {e}"
        ) from e
```

By doing so you don't need to check for the `is_dir()` inside `is_model_safetensors`.

leseb reviewed (changes requested):
I agree with @danmcp that this is a good candidate for unit tests. Please add them. This is a step in the right direction and an improvement over what we currently have. The validation could be enhanced further by:

- Verifying that the JSON files are valid.
- Reading the header of the safetensors files to confirm they are indeed safetensors files.

We need to find the right balance in how thorough we want the validation to be. This can be done in another PR though. Thanks!

jaideepr97 commented:
@danmcp @leseb 
I've rebased on top of sebastian's recently merged PR and addressed the comments for the most part
The one thing I haven't added yet is a concrete way to detect safetensors files (via file header etc) since I haven't been able to find an easy way to do that similar to how GGUF files have magic numbers. If either of you have pointers for me on that specific aspect that would be much appreciated!

leseb commented on src/instructlab/model/backends/backends.py:
```diff
@@ -86,6 +88,45 @@ def shutdown(self):
+    Returns:
+        bool: True if the model is a safetensors model, False otherwise.
+    """
+    is_valid_safetensors_model = True
```
It's uncommon to default to True. If there is a bug somewhere and we don't see it immediately this will return True. Please default to False.

jaideepr97 replied:
makes sense, updated!

leseb commented on src/instructlab/model/backends/backends.py:
```diff
@@ -117,29 +158,29 @@ def determine_backend(model_path: pathlib.Path) -> Tuple[str, str]:
         if sys.platform == "linux":
             logger.debug(
-                f"Model is a directory and system is Linux, using {VLLM} backend."
+                f"Model is a huggingface safetensors model and system is Linux, using {VLLM} backend."
```
```suggestion
                f"Model is huggingface safetensors and system is Linux, using {VLLM} backend."
```

leseb replied:
"model" is a bit redundant here.

jaideepr97 replied:
updated

leseb commented on src/instructlab/model/backends/backends.py:
```diff
@@ -117,29 +158,29 @@ def determine_backend(model_path: pathlib.Path) -> Tuple[str, str]:
-            )
-        # If the model is a directory, it's a VLLM model - it's kinda weak, but it's a start
+    
+    if model_path.is_dir() and is_model_safetensors(model_path):
```
Given that `is_model_safetensors` can raise something I think this needs to go in a try-expect block.

jaideepr97 replied:
added, thanks!

leseb commented on src/instructlab/model/backends/backends.py:
```diff
@@ -117,29 +158,29 @@ def determine_backend(model_path: pathlib.Path) -> Tuple[str, str]:
                 "model path is a directory containing huggingface safetensors files and running on Linux.",
             )
+        else:
+            raise ValueError(
```
This is probably the `except` I mentioned above.

jaideepr97 replied:
done

jaideepr97 replied:
on second thought, i dont think this should be an exception, this should still be an else block as we want it to get invoked when nothing has gone wrong, we have just detected that we are running on non-linux and we should call that out

I will add a separate exception handler as well

leseb reviewed (changes requested):
I don't think we need to have "real" files in the test_data dir. For instance, tests/testdata/valid_safetensors_model_dir/tokenizer.json is 99K lines... It's a bit extreme to have that file in here. The implementation just validates the presence of the files and their format (json), not the content. So mocking is sufficient.

Thanks!

jaideepr97 commented:
@leseb I removed the really large file, but kept some smaller json files because we are doing some basic json validation 
hope that is acceptable

leseb reviewed (commented):
I'm happy with the direction this is taking, I just need some time to go through the unit tests. In the meantime, can you look at my latest suggestion? Thanks

jaideepr97 commented on src/instructlab/utils.py:
```diff
@@ -350,6 +350,24 @@ def retrieve(uri: str) -> Resource:
     return errors
 
 
+def validate_json_file(json_filepath: Path) -> bool:
```
It seemed like a general utility function that we can write once and reuse wherever we need to validate json files, instead of duplicating that check everywhere 
I'm happy to remove it if you feel strongly though

leseb replied:
If the function was doing something else than just validating the file, like checking for fields etc I would say it makes sense to have it. But right now, I'm sure it's helping much but maybe others think differently...

tiran replied:
I strongly advise against unstructured JSON or unstructured data in general. It's a source of bugs and pain. All JSON or YAML documents have to be defined by a schema and validated..

jaideepr97 replied:
@tiran sorry, not quite sure I follow what you're suggesting here
If you're suggesting that we need more strict JSON validation, maybe this function can be expanded to do schema validation in a future PR (and maybe there is value in keeping it here then?)

alimaredia replied:
I think for the scope of this PR we remove safetensor model json file validation right now and make it a TODO along with verifying .safetensors files are valid as well.

leseb replied:
Discussed offline with @tiran and we would like to remove this function as suggested previously and simply do try-except in place. Thanks!

jaideepr97 replied:
removed, thanks!

tiran commented on tests/test_backends.py:
```diff
@@ -47,33 +48,76 @@ def test_get_backend_auto_detection_success_gguf(
-        backends.get(tmp_path, None)
-    assert "Cannot determine which backend to use" in str(exc_info.value)
+# this tests both cases where a valid and invalid safetensors model directory is supplied
+@pytest.mark.parametrize(
+    "safetensors_model_path,expected",
+    [
+        (
+            pathlib.Path(
+                os.path.join(
+                    pathlib.Path(__file__).parent.resolve(),
+                    "testdata/valid_safetensors_model_dir",
+                )
+            ),
+            True,
+        ),
+        (
+            pathlib.Path(
+                os.path.join(
+                    pathlib.Path(__file__).parent.resolve(),
+                    "testdata/invalid_safetensors_model_dir",
+                )
+            ),
+            False,
+        ),
+    ],
+)
+def test_is_model_safetensors_valid(safetensors_model_path, expected):
```
Change the parameters to "safetensor_dir,expected", [("valid_safetensors_model_dir", True), ...] and then use the `testdata_path` fixture to get the path.
Also add type hints to all parameters

```
def test_is_model_safetensors_valid(safetensors_dir: str, expected: bool, testdata_path: pathlib.Path) ->:
    safetensors_model_path = testdata_path / safetensors_dir
```

jaideepr97 replied:
updated, thanks!

alimaredia commented:
If the .json file validation is removed from this PR, I will be ready to go ahead an approve it.

@leseb do you think follow up functional tests are needed for this PR or will unit the unit tests tests do? It seems like there's 3 follow up work items needed that someone can take up:
1. `.safetensors` file validation
2. `.json` file validation for a .safetensors model
3. functional tests

leseb reviewed (commented):
It looks like we don't validate any of the fields in JSON right now, so I'd like to remove all the fake test files and just do a mocking of a dummy `{}` file in the test. Thanks

leseb commented:
> If the .json file validation is removed from this PR, I will be ready to go ahead an approve it.
> 
> @leseb do you think follow up functional tests are needed for this PR or will unit the unit tests tests do? It seems like there's 3 follow up work items needed that someone can take up:
> 
> 1. `.safetensors` file validation

It depends how far we want to push the validation, really.

> 2. `.json` file validation for a .safetensors model

It's already in this PR.

> 3. functional tests
We have all the unit tests we need for that, so the validation will happen naturally when the CI execute a runner that has vLLM, not specific func test required here.

leseb commented:
TODO before we can merge:

- [ ] Remove all the dummy test files and use a simple json mock
- [ ] squash all commits into one

Thanks @jaideepr97

jaideepr97 commented:
@leseb done

alimaredia commented on tests/test_backends.py:
```diff
@@ -49,33 +52,99 @@ def test_get_backend_auto_detection_success_gguf(
+        f.write("")
+
+    backend = backends.get(safetensors_model_path, None)
+    assert backend == expected
```
Is expected here supposed to be a `bool` or a `str`?  Isn't the value of backend here supposed to be "vllm"?

jaideepr97 replied:
it is "vllm", if you look at line 104

jaideepr97 replied:
fixed, thanks

alimaredia commented on tests/test_backends.py:
```diff
@@ -49,33 +52,99 @@ def test_get_backend_auto_detection_success_gguf(
+            safetensors_model_path / "tokenizer.model", "a+", encoding="UTF-8"
+        ) as f:
+            f.write("")
+
```
The safetensors model path creation should get moved to a function. I see it being repeated in the first two new tests.

jaideepr97 replied:
done

leseb commented on tests/test_backends.py:
```diff
@@ -49,33 +73,71 @@ def test_get_backend_auto_detection_success_gguf(
+        ),
+    ],
+)
+def test_is_model_safetensors_valid(
```
It looks like the tests are not using the `tmp_path` fixture from pytest. Please re-add that and look at other tests for reference. Also, if you do that you don't need to call `shutil.rmtree()` anymore. Thanks!

jaideepr97 replied:
fixed, thanks!

leseb reviewed (commented):
Please use the tmp_path fixture in the test, then LGTM :D Thanks!

tiran commented on src/instructlab/model/backends/backends.py:
```diff
@@ -111,6 +114,48 @@ def safe_close_all(resources: typing.Iterable[Closeable]):
+        bool: True if the model is a safetensors model, False otherwise.
+    """
+
+    try:
```
It's bad style to wrap so much code in a try/except block. It can unrelated bugs in your code.

tiran commented on src/instructlab/model/backends/backends.py:
```diff
@@ -111,6 +114,48 @@ def safe_close_all(resources: typing.Iterable[Closeable]):
+    """
+
+    try:
+        required_files = frozenset(
+            {
+                "config.json",
+                "*.safetensors",
+                "tokenizer.json",
+                "tokenizer.model",
+                "tokenizer_config.json",
+            }
+        )
```
There is no need to create a frozen set here. The variable never leaves the scope of the function.

tiran commented on src/instructlab/model/backends/backends.py:
```diff
@@ -111,6 +114,48 @@ def safe_close_all(resources: typing.Iterable[Closeable]):
+            }
+        )
+
+        if not all(glob.glob(os.path.join(model_path, re)) for re in required_files):
```
`model_path` is a pathlib object, use its powers and features.

tiran reviewed (commented):
Here is how I would write the function (untested)

```python
def is_model_safetensors(model_path: pathlib.Path) -> bool:
    """Check if model_path is a valid safe tensors directory

    Check if provided path to model represents directory containing a safetensors representation
    of a model. Directory must contain a specific set of files to qualify as a safetensors model directory
    Args:
        model_path (Path): The path to the model directory
    Returns:
        bool: True if the model is a safetensors model, False otherwise.
    """
    try:
        files = list(model_path.iterdir())
    except (FileNotFoundError, NotADirectoryError, PermissionError) as e:
        logger.debug("Failed to read directory: %s", e)
        return False

    if not any(file.suffix == ".safetensors" for file in files):
        logger.debug("'%s' has no *.safetensors files", model_path)
        return False

    basenames = {file.name for file in files}
    requires_files = {
        "config.json",
        "tokenizer.json",
        "tokenizer.model",
        "tokenizer_config.json",
    }
    diff = requires_files.difference(basenames)
    if diff:
        logger.debug("'%s' is missing %s", model_path, diff)
        return False

    for file in model_path.glob("*.json"):
        try:
            with file.open(encoding="utf-8") as f:
                json.load(f)
        except (PermissionError, json.JSONDecoder) as e:
            logger.debug("'%s' is not a valid JSON file: e", file, e)
            return False

    # TODO: add check for safetensors file header (?)
    return True
```

jaideepr97 commented:
thanks @tiran, updated the function as per your suggestions 
cc @leseb
